lib/eig_iram.cpp

```diff
@@ -0,0 +1,567 @@
+    // to the j step factorization.            
+    // Use Classical Gram Schmidt and compute: 
+    // w_{j} <-  V_{j}^dag * M * v_{j}      
+    // r_{j} <-  M * v_{j} - V_{j} * w_{j}
+
+    //H_{j,i}_j = v_i^dag * r
+    std::vector<Complex> tmp(j+1);
+    std::vector<ColorSpinorField *> v_;
+    v_.reserve(j+1);
+    for (int i = 0; i < j+1; i++) { v_.push_back(v[i]); }    
+    blas::cDotProduct(tmp.data(), v_, r);
+
+    // Orthogonalise r_{j} against V_{j}.
+    //r = r - H_{j,i} * v_j 
+    for (int i = 0; i < j+1; i++) tmp[i] *= -1.0;    
+    blas::caxpy(tmp.data(), v_, r);
+    for (int i = 0; i < j+1; i++) upperHess[i][j] = -1.0*tmp[i];
+    
+    // Re-orthogonalization / Iterative refinement phase 
+    // Maximum 100 tries.                                
+
+    // s      = V_{j}^T * B * r_{j}
+    // r_{j}  = r_{j} - V_{j}*s    
+    // alphaj = alphaj + s_{j}     
+
+    // The stopping criteria used for iterative refinement is    
+    // discussed in Parlett's book SEP, page 107 and in Gragg &  
+    // Reichel ACM TOMS paper; Algorithm 686, Dec. 1990.         
+    // Determine if we need to correct the residual. The goal is 
+    // to enforce ||v(:,1:j)^T * r_{j}|| .le. eps * || r_{j} ||  
+    // The following test determines whether the sine of the     
+    // angle between  OP*x and the computed residual is less     
+    // than or equal to 0.717.
+
+    int orth_iter = 0;
+    int orth_iter_max = 100;
+    beta = sqrt(blas::norm2(*r[0]));
+    while(beta < 0.717*beta_pre && orth_iter < orth_iter_max) {
+
+      if (getVerbosity() >= QUDA_DEBUG_VERBOSE) {
+	printfQuda("beta = %e > 0.717*beta_pre = %e: Reorthogonalise at step %d, iter %d\n", beta, 0.717*beta_pre, j, orth_iter);
+      }
+
+      beta_pre = beta;
+      
+      // Compute the correction to the residual:
+      // r_{j} = r_{j} - V_{j} * r_{j}
+      // and adjust for the correction in the
+      // upper Hessenberg matrix.
+      blas::cDotProduct(tmp.data(), v_, r);
+      for (int i = 0; i < j+1; i++) tmp[i] *= -1.0;    
+      blas::caxpy(tmp.data(), v_, r);    
+      for (int i = 0; i < j+1; i++) upperHess[i][j] -= tmp[i];
+      
+      beta = sqrt(blas::norm2(*r[0]));
+      orth_iter++;
+    }
+
+    if(orth_iter == orth_iter_max) {
+      errorQuda("Unable to orthonormalise r");
+    }
+  }
+  
+  void IRAM::rotateBasis(std::vector<ColorSpinorField *> &kSpace, int keep)
+  {
+    // Multi-BLAS friendly array to store the part of the rotation matrix
+    Complex *Qmat_keep = (Complex *)safe_malloc((n_kr * keep) * sizeof(Complex));
+    for (int j = 0; j < n_kr; j++)       
+      for (int i = 0; i < keep; i++) { Qmat_keep[j * keep + i] = Qmat[j][i]; }
+    
+    rotateVecsComplex(kSpace, Qmat_keep, n_kr, n_kr, keep, profile);
+    
+    host_free(Qmat_keep);
+  }
+
+  void IRAM::reorder(std::vector<ColorSpinorField *> &kSpace, std::vector<Complex> &evals, const QudaEigSpectrumType spec_type)
+  {
+    int i=0;
+    switch(spec_type) {
```

Comment:
For code cleanliness, you can pull off something similar to `void EigenSolver::sortArrays(QudaEigSpectrumType spec_type, int n, std::vector<Complex> &x, std::vector<Complex> &y)` here using a `std::tuple` instead of an `std::pair`.